pennylane/math/quantum.py

```diff
@@ -164,6 +169,167 @@ def marginal_prob(prob, axis):
     return np.flatten(prob)
 
 
+def _density_matrix_from_matrix(density_matrix, wires, check_state=None):
+    """Compute the density matrix from a state vector.
+
+    Args:
+        density_matrix (tensor_like): 1D tensor state vector. This tensor should be of size ``(2**N,)`` for some
+            integer number of wires``N``.
+        wires (list(int)): List of wires (int) in the subsystem.
+        check_state (bool): If True, the function will check the state validity (shape and norm).
+
+    Returns:
+        tensor_like: Density matrix of size ``(2**len(wires), 2**len(wires))``
+
+    **Example**
+
+    >>> x = np.array([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])
+    >>> _density_matrix_from_matrix(x, wires=[0])
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]]
+
+    >>> y = [[0.5, 0, 0.5, 0], [0, 0, 0, 0], [0.5, 0, 0.5, 0], [0, 0, 0, 0]]
+    >>> _density_matrix_from_matrix(y, wires=[0])
+    [[0.5+0.j 0.5+0.j]
+     [0.5+0.j 0.5+0.j]]
+
+    >>> _density_matrix_from_matrix(y, wires=[1])
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]]
+
+    >>> z = tf.Variable([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], dtype=tf.complex128)
+    >>> _density_matrix_from_matrix(x, wires=[1])
+    tf.Tensor(
+    [[1.+0.j 0.+0.j]
+     [0.+0.j 0.+0.j]], shape=(2, 2), dtype=complex128)
+
+
+    """
+    density_matrix = cast(density_matrix, dtype="complex128")
+    shape = density_matrix.shape[0]
+    num_wires = int(np.log2(shape))
+
+    if check_state:
+        # Check format
+        if (
+            len(density_matrix.shape) != 2
+            or density_matrix.shape[0] != density_matrix.shape[1]
+            or np.ceil(np.log2(shape)) != np.floor(np.log2(shape))
+        ):
+            raise ValueError("Density matrix must be of shape (2**N, 2**N).")
+        # Check trace
+        trace = np.trace(density_matrix)
+        if not is_abstract(trace):
+            if not allclose(trace, 1.0, atol=1e-10):
+                raise ValueError("The trace of the density matrix should be one.")

```

Comment:
Question: Since we're checking the shape and trace of the density matrix, should we also check if it's positive semi-definite? A matrix such as `np.array([[2, 0], [0, -1]])` would pass all the current checks.